utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -457,130 +527,153 @@ pub enum State {
+	///
+	/// This will override the code as it sees fit based on [`SharedParams::Runtime`]. It will also
+	/// check the spec-version and name.
+	pub(crate) async fn into_ext<Block: BlockT + DeserializeOwned, HostFns: HostFunctions>(
+		&self,
+		shared: &SharedParams,
+		config: &Configuration,
+		executor: &WasmExecutor<HostFns>,
+	) -> sc_cli::Result<RemoteExternalities<Block>>
 	where
 		Block::Hash: FromStr,
+		Block::Header: DeserializeOwned,
+		Block::Hash: DeserializeOwned,
 		<Block::Hash as FromStr>::Err: Debug,
 	{
-		Ok(match self {
+		let builder = match self {
 			State::Snap { snapshot_path } =>
 				Builder::<Block>::new().mode(Mode::Offline(OfflineConfig {
 					state_snapshot: SnapshotConfig::new(snapshot_path),
 				})),
-			State::Live { snapshot_path, pallet, uri, at, child_tree } => {
+			State::Live(LiveState { pallet, uri, at, child_tree, threads }) => {
 				let at = match at {
 					Some(at_str) => Some(hash_of::<Block>(at_str)?),
 					None => None,
 				};
-				let mut builder = Builder::<Block>::new()
-					.mode(Mode::Online(OnlineConfig {
-						transport: uri.to_owned().into(),
-						state_snapshot: snapshot_path.as_ref().map(SnapshotConfig::new),
-						pallets: pallet.clone(),
-						scrape_children: true,
-						at,
-					}))
-					.inject_hashed_key(
-						&[twox_128(b"System"), twox_128(b"LastRuntimeUpgrade")].concat(),
-					);
-				if *child_tree {
-					builder = builder.inject_default_child_tree_prefix();
-				}
-				builder
+				Builder::<Block>::new().mode(Mode::Online(OnlineConfig {
+					at,
+					transport: uri.to_owned().into(),
+					state_snapshot: None,
+					pallets: pallet.clone(),
+					child_trie: *child_tree,
+					hashed_keys: vec![
+						// we always download the code, but we almost always won't use it, based on
+						// `Runtime`.
+						well_known_keys::CODE.to_vec(),
+						// we will always download this key, since it helps detect if we should do
+						// runtime migration or not.
+						[twox_128(b"System"), twox_128(b"LastRuntimeUpgrade")].concat(),
+					],
+					hashed_prefixes: vec![],
+					threads: *threads,
+				}))
 			},
-		})
-	}
+		};
+
+		// possibly overwrite the state version, should hardly be needed.
+		let builder = if let Some(state_version) = shared.overwrite_state_version {
+			log::warn!(
+				target: LOG_TARGET,
+				"overwriting state version to {:?}, you better know what you are doing.",
+				state_version
+			);
+			builder.overwrite_state_version(state_version)
+		} else {
+			builder
+		};
+
+		let mut ext = builder.build().await?;
+		let original_code = ext
+			.execute_with(|| sp_io::storage::get(well_known_keys::CODE))
+			.expect("':CODE:' is always downloaded in try-runtime-cli; qed");
+
+		// then, we replace the code based on what the CLI wishes.
+		let maybe_code_to_overwrite = match shared.runtime {
+			Runtime::Local => Some(
+				config
+					.chain_spec
+					.build_storage()
+					.unwrap()
```

Comment:
question: this is still suboptimal, right? especially building the whole unnecessary storage; tbh I don't see any gain of keeping `Runtime::Local` when you can just point to a compiled wasm blob

Reply:
yeah, let me try the wasm blob approach and if it works, I will just remove it.